packages/core/src/browser/label-provider.ts

```diff
@@ -0,0 +1,116 @@
+/*
+* Copyright (C) 2017 TypeFox and others.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { inject, injectable, named } from "inversify";
+import * as fileIcons from "file-icons-js";
+import URI from "../common/uri";
+import { ContributionProvider } from '../common/contribution-provider';
+import { Prioritizeable } from '../common/types';
+
+import "file-icons-js/css/style.css";
+
+export const LabelProviderContribution = Symbol('LabelProviderContribution');
+export interface LabelProviderContribution {
+
+    /**
+     * whether this contribution can handle the given element and with what priority.
+     * All contributions are ordered by the returned number if greater than zero. The highest number wins.
+     * If two or more contributions return the same positive number one of those will be used. It is undefined which one.
+     */
+    canHandle(element: object): number;
+
+    /**
+     * returns an icon class for the given element.
+     */
+    getIcon?(element: object): string;
+
+    /**
+     * returns a short name for the given element.
+     */
+    getName?(element: object): string;
+
+    /**
+     * returns a long name for the given element.
+     */
+    getLongName?(element: object): string;
+
+}
+
+@injectable()
+export class DefaultUriLabelProviderContribution implements LabelProviderContribution {
+
+    canHandle(uri: object) {
+        if (uri instanceof URI) {
+            return 1;
+        }
+        return 0;
+    }
+
+    getIcon(uri: URI): string {
+        const iconClass = fileIcons.getClass(uri.path.toString());
+        if (!iconClass) {
+            if (uri.displayName.indexOf('.') === -1) {
+                return 'fa fa-folder';
+            } else {
+                return 'fa fa-file';
+            }
+        }
+        return iconClass;
+    }
+
+    getName(uri: URI): string {
+        return uri.displayName;
```

Comment:
I wonder whether we should inline uri.displayName here to force clients to use `LabelProvider`

Reply:
inline?

Reply:
I think it is ok to have a way to have a short `toString()` directly on the URI. For all code that might not have access to injected components. At least I would prefer to not add that to this PR, because it will involve a lot of changes.

Reply:
ok

Reply:
I asked because I had added `displayName` to `URI` because of not having a better place, now `LabelProvider` looks like the proper place. But it can be reconsidered in another PR.